tokio/src/task/coop/mod.rs

```diff
@@ -107,8 +167,93 @@ fn with_budget<R>(budget: Budget, f: impl FnOnce() -> R) -> R {
     f()
 }
 
+/// Returns `true` if there is still budget left on the task.
+///
+/// Futures created by the tokio library functions are budget-aware and yield when there is no more
+/// budget left, but not all futures will be budget-aware. Consider future `A` that polls two inner
+/// futures `B` and `C`, and returns `Poll::Ready` when one of them is ready. If both inner futures
+/// were budget-aware, at some point the budget would be depleted which would cause both futures to
+/// return `Poll::Pending` resulting in `A` returning `Poll::Pending` as well. Yielding all the way
+/// back to the runtime, the budget would be reset, and `B` and `C` could make progress again.
+/// Now let's consider `B` is budget-aware, but `C` is not and is always ready. The budget will be
+/// depleted as before, but now since `C` always returns `Poll::Ready`, `A` will always return
+/// `Poll::Ready` as well. This way, `A` will not yield back to the runtime which will keep the budget
+/// depleted and `B` not making progress.
```

Comment:
This explanation is confusing. And more generally, this seems like a niche use-case. Are there not more common use-cases for this that we can explain instead?

E.g. maybe just have this example:
```rs
if !has_budget_remaining() {
    tokio::task::yield_now().await;
}
```
and say that this is similar to `consume_budget`, but doesn't decrement the budget when non-zero.

Reply:
You're right. I'll update the docs as suggested.